### ISSUE TITLE
Tensor manipulation routines

### ISSUE THREAD

**Issue by rsokl** [rsokl/MyGrad#90]
The following [tensor manipulation routines](https://docs.scipy.org/doc/numpy-1.15.1/reference/routines.array-manipulation.html) should be implemented:

- [x] ravel
- [x] Tensor.flatten
- [x] expand_dims
- [x] broadcast_to

**Comment by IanCoolidge0** [comment]
The backward passes for these should just perform the corresponding operation on the gradient, correct?

**Comment by rsokl** [comment]
The back-prop needs to "undo" each manipulation.

For example, `ravel(x)` flattens `x`. Thus the the gradient getting back-propped to this operation must be a flat tensor (the gradient must always have the same shape as its associated tensor). Backpropagating the gradient to `x` entails reshaping the gradient so that its elements correspond with the elements of `x`.

[Further reference](https://github.com/rsokl/MyGrad/blob/master/mygrad/tensor_manip/array_shape/ops.py#L21)

`broadcast_to` is a bit more complicated - you need to sum-reduce the elements of the broadcasted gradient back down to the original shape of the tensor. [This utility](https://github.com/rsokl/MyGrad/blob/master/mygrad/_utils.py) already has this capability. Feel free to skip this one if you want, as I really should show you the derivation for its gradient.

**Comment by IanCoolidge0** [comment]
That makes sense, ravel and flatten should be easy enough in that case. I'll take a look at the `broadcast_to` calculation you linked since I guess that would apply to expand dims also.

**Comment by rsokl** [comment]
You are correct, however `expand_dims` can be undone by `reshape` as well, I believe. 

Given that it introduces ["placeholder" dimensions](https://www.pythonlikeyoumeanit.com/Module3_IntroducingNumpy/Broadcasting.html#Size-1-Axes-&-The-newaxis-Object), summing over those axes will not have any accumulating effect on the image. Reshaping can effectively "squeeze out" these dimensions.

```python
# a shape (1, 1, 2, 1) array
>>> np.array([[[[3], [4]]]]). reshape(2)
array([3, 4]
```

**Comment by rsokl** [comment]
Closed with #103
